### PR TITLE
Adding support for AWS S3 binary caching

### PR DESCRIPTION
**Pull request by dave-juicelabs**

### CODE REVIEW

**Comment by dave-juicelabs** [comment]
While this is being reviewed, is there a way for me to use this in my own personal builds? Since vcpkg now downloads a built version, I am unsure of how I would be able to integrate this.

**Comment by mathisloge** [comment]
> While this is being reviewed, is there a way for me to use this in my own personal builds? Since vcpkg now downloads a built version, I am unsure of how I would be able to integrate this.

you can just replace the downloaded version of vcpkg with your built version

File: src/vcpkg/binarycaching.cpp [hunk 1]
```diff
@@ -1120,6 +1120,186 @@ namespace
...
+    {
+        Command cmd;
+        cmd.string_arg("aws").string_arg("s3").string_arg("cp").string_arg(aws_object).path_arg(archive);
+        const auto out = cmd_execute_and_capture_output(cmd);
+        if (out.exit_code == 0)
+        {
+            return true;
+        }
+
+        print2(Color::warning, "aws failed to download with exit code: ", out.exit_code, '\n', out.output);
```
Comment (strega-nil-ms) [inline src/vcpkg/binarycaching.cpp:1156]: This should be using the new localized output; at the top of the file:

```cxx
DECLARE_AND_REGISTER_MESSAGE(AwsFailedToDownload, (msg::value, msg::output), "aws failed to download with exit code: {value}\n{output}");
```

(then add a new `DECLARE_MSG_ARG` in messages.h for `output`)

Then, replace this with:

```suggestion
        msg::println(Color::warning, msgAwsFailedToDownload, msg::value = out.exit_code, msg::output = out.output);
```

File: src/vcpkg/binarycaching.cpp [hunk 2]
```diff
@@ -1120,6 +1120,186 @@ namespace
...
+
+                    clean_prepare_dir(fs, paths.package_dir(action.spec));
+                    url_paths.emplace_back(make_aws_path(prefix, *abi),
+                                           make_temp_archive_path(paths.buildtrees(), action.spec));
+                    url_indices.push_back(idx);
+                }
+
+                if (url_paths.empty()) break;
+
+                print2("Attempting to fetch ", url_paths.size(), " packages from AWS.\n");
```
Comment (strega-nil-ms) [inline src/vcpkg/binarycaching.cpp:1203]: same as above - replace with localized output.

File: src/vcpkg/tools.cpp [hunk 3]
```diff
@@ -567,6 +567,35 @@ gsutil version: 4.58
         }
     };
 
+    struct AWSCliProvider : ToolProvider
```
Comment (strega-nil-ms) [inline src/vcpkg/tools.cpp:570]:
```suggestion
    struct AwsCliProvider : ToolProvider
```

Acronyms should be treated as a single word for the purposes of camel case, like in [rust].

[rust]: https://rust-lang.github.io/api-guidelines/naming.html

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
Some minor changes, but otherwise, this looks great :)

**Comment by strega-nil-ms** [comment]
cc @ras0219-msft as the owner of binary caching.

**Comment by dave-juicelabs** [comment]
Ahh, yeah that looks a lot more reasonable.

**Comment by Pospelove** [comment]
Can't wait for AWS S3 support :)

**Review by ras0219-msft — APPROVED** [review]
I believe this also will require adding the AWS CLI to `vcpkgTools.xml` in the vcpkg repo.

Otherwise, LGTM. Thanks for the PR!

**Comment by strega-nil-ms** [comment]
thanks y'all!

**Pull request merged**